### PR TITLE
Resolves #1824: Use different analyzers for different fields in one L…

### LINKED ISSUE #1824: Use different analyzers for different fields in one Lucene index, and enable exclusion of fields for auto-complete

### PR DESCRIPTION
**Pull request by tian-yizuo**
…ucene index, and enable exclusion of fields for auto-complete

### CODE REVIEW

**Review by MMcM — APPROVED** [review]
Although I think the scheme discussed in the meeting of having per-field option _names_ with a common prefix (followed by the field name) would be more straightforward, it looks like this ought to work okay.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java [hunk 1]
```diff
@@ -60,16 +60,32 @@ public class LuceneIndexOptions {
...
+     * The format of the value should be "fieldName1/analyzerName1/fieldName2/analyzerName2...".
+     */
+    public static final String LUCENE_ANALYZER_NAME_PER_FIELD_OPTION = "luceneAnalyzerNamePerField";
+    /**
+     * The name of the Lucene analyzer to use for auto-complete query.
      */
     public static final String AUTO_COMPLETE_ANALYZER_NAME_OPTION = "autoCompleteAnalyzerName";
+    /**
+     * The override mapping from fields to Lucene analyzers for auto-complete query, if they want to use different analyzers than the default one.
+     * The format of the value should be "fieldName1/analyzerName1/fieldName2/analyzerName2...", with keys and values splitted by "/".
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java:75]:
```suggestion
     * The format of the value should be "fieldName1/analyzerName1/fieldName2/analyzerName2...", with keys and values split by "/".
```

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java [hunk 2]
```diff
@@ -60,16 +60,32 @@ public class LuceneIndexOptions {
...
+     */
+    public static final String AUTO_COMPLETE_ANALYZER_NAME_PER_FIELD_OPTION = "autoCompleteAnalyzerNamePerField";
     /**
      * The name of the synonym set to use in Lucene.
      */
     public static final String TEXT_SYNONYM_SET_NAME_OPTION = "textSynonymSetName";
 
+    /**
+     * A list of fields to be excluded for auto-complete query.
+     * The format of the value should be "fieldName1/fieldName2/...", with field names splitted by "/".
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexOptions.java:85]:
```suggestion
     * The format of the value should be "fieldName1/fieldName2/...", with field names split by "/".
```

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Provide a combination of analyzers for multiple fields of one Lucene index.
+ * The combinations of analyzers for index time and query time are potentially different.
+ * The default analyzer chooser is used for all fields of one Lucene index except the fields which has overrides in the analyzer chooser per field mapping.
+ */
+public class LuceneAnalyzerCombinationProvider {
```
[the remark this answers is not in the document]
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java:39]: The ```AnalyzerChooser``` is designed for the provision of an Lucene analyzer given a text, so it is per Lucene analyzer. The only reason we need it for now, is for the capability needed to use different one for different languages. Whereas the ```LuceneAnalyzerCombinationProvider``` is more like a single object to provide a pair of analyzers for index time and query time based on the options config, because we want to make is simple for clients to configure for the analyzers to use, without worrying about which to use on index time and which to use on query time.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java [hunk 4]
```diff
@@ -86,17 +88,41 @@ private LuceneAnalyzerRegistryImpl() {
...
+        Pair<AnalyzerChooser, AnalyzerChooser> defaultAnalyzerChooserPair = getAnalyzerChooser(index, defaultAnalyzerName, type);
+
+        if (analyzerPerFieldName == null) {
+            return new LuceneAnalyzerCombinationProvider(defaultAnalyzerChooserPair.getLeft(), defaultAnalyzerChooserPair.getRight(),
+                    null, null);
+        }
+
+        Map<String, AnalyzerChooser> indexAnalyzerChooserPerFieldOverride = new TreeMap<>();
+        Map<String, AnalyzerChooser> queryAnalyzerChooserPerFieldOverride = new TreeMap<>();
+        String[] elements = analyzerPerFieldName.split("/");
```
[the remark this answers is not in the document]
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:103]: @alecgrieser yeah it sounds better to use different delineators for that between key-value pairs and that between keys and values. So can I assume no field names will have ```:``` or ```,```?

[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:103]: I don't know how valid an assumption that is; I think `:` and `,` are better delineator characters for maps, but I don't know whether they're currently allowed in field names (or analyzer names) or not. That's one of the reasons that for, say, the "replaced by" index name options, rather than coming up with a scheme by which the index names could be put together in a single option, we went with instead a single option _prefix_, and then collected all IndexOptions that began with that prefix to produce the list.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerCombinationProvider.java:39]: I think I like the abstraction, for the most part, and that it makes sense to encapsulate index-and-query-time-analyzers into an object other than a pair. However, once you have that abstraction, I'm not sure it makes sense to combine that with the behavior that constructs the per-field Lucene analyzers.

I also think the abstraction is fine for now, though, and something we could iterate on later if we really want to

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java [hunk 5]
```diff
@@ -44,5 +46,49 @@ public void validate(@Nonnull MetaDataValidator metaDataValidator) {
...
+                    || analyzerNamePerFieldOption.startsWith(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS)
+                    || analyzerNamePerFieldOption.endsWith(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE)
+                    || analyzerNamePerFieldOption.endsWith(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS)) {
+                throw new MetaDataException("Index " + index.getName() + " has invalid " + optionKey + ": " + analyzerNamePerFieldOption);
+            }
+            String[] keyValuePairs = analyzerNamePerFieldOption.split(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS);
+            for (String keyValue : keyValuePairs) {
+                if (keyValue.indexOf(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE) < 1
+                        || keyValue.indexOf(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE) > keyValue.length() - 2
+                        || keyValue.indexOf(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE) != keyValue.lastIndexOf(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE)) {
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:74]: This isn't exactly performance critical code, but this recomputes `keyValue.indexOf(':')` multiple times. Ideally, this would compute the value once (for each `keyValue`) and store it in a local variable.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java [hunk 6]
```diff
@@ -44,5 +46,49 @@ public void validate(@Nonnull MetaDataValidator metaDataValidator) {
...
+                    throw new MetaDataException("Index " + index.getName() + " has invalid " + optionKey + ": " + analyzerNamePerFieldOption);
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:75]:
```suggestion
                    throw new MetaDataException("Index " + index.getName() + " has invalid option value " + optionKey + ": " + analyzerNamePerFieldOption);
```

Likewise for the other `MetaDataExceptions` that are thrown in this class

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java [hunk 7]
```diff
@@ -44,5 +46,49 @@ public void validate(@Nonnull MetaDataValidator metaDataValidator) {
...
+    private static void validateAnalyzerNamePerFieldOption(@Nonnull String optionKey, @Nonnull Index index) {
+        String analyzerNamePerFieldOption = index.getOption(optionKey);
+        if (analyzerNamePerFieldOption != null) {
+            analyzerNamePerFieldOption = analyzerNamePerFieldOption.strip();
+            if (analyzerNamePerFieldOption.startsWith(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE)
+                    || analyzerNamePerFieldOption.startsWith(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS)
+                    || analyzerNamePerFieldOption.endsWith(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE)
+                    || analyzerNamePerFieldOption.endsWith(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS)) {
+                throw new MetaDataException("Index " + index.getName() + " has invalid " + optionKey + ": " + analyzerNamePerFieldOption);
+            }
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:69]: I think these checks are all extraneous, as they'd all get validated by the later check of each key-value pair.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java [hunk 8]
```diff
@@ -44,5 +46,49 @@ public void validate(@Nonnull MetaDataValidator metaDataValidator) {
...
+    public static void validateIndexOptions(@Nonnull Index index) {
+        validateAnalyzerNamePerFieldOption(LuceneIndexOptions.LUCENE_ANALYZER_NAME_PER_FIELD_OPTION, index);
+        validateAnalyzerNamePerFieldOption(LuceneIndexOptions.AUTO_COMPLETE_ANALYZER_NAME_PER_FIELD_OPTION, index);
+        validateAutoCompleteExcludedFields(index);
+    }
+
+    private static void validateAnalyzerNamePerFieldOption(@Nonnull String optionKey, @Nonnull Index index) {
+        String analyzerNamePerFieldOption = index.getOption(optionKey);
+        if (analyzerNamePerFieldOption != null) {
+            analyzerNamePerFieldOption = analyzerNamePerFieldOption.strip();
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:63]: I don't think the way that `strip()` is handled here is quite right, because `strip` is not called when the index options are actually evaluated later

**Review by alecgrieser — CHANGES_REQUESTED** [review]
In my original review, I also proposed stripping out whitespace from the field and analyzer name options, but that wasn't done here, except somewhat during validation. I think it would be reasonable to either go with my initial suggestion (and strip out whitespace for each field and analyzer name) or to have tighter validation that required that there not be whitespace between elements.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java (hunk 8 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:63]: Changed it to also call the ```strip()``` when actually evaluating the index options.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java (hunk 7 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Good catch, removed these

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java (hunk 6 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Fixed it

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java (hunk 5 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Fixed it

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java [hunk 9]
```diff
@@ -1,2 +1,88 @@
...
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+import static com.apple.foundationdb.record.metadata.Key.Expressions.concat;
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
+import static com.apple.foundationdb.record.metadata.Key.Expressions.function;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class LuceneIndexValidatorTest {
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:36]: I'm not sure this belongs in the `LuceneIndexValidatorTest`, but we probably want positive test cases somewhere as well. I...suppose this doesn't need to block this PR, though

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java [hunk 10]
```diff
@@ -100,11 +100,13 @@ public LuceneAnalyzerCombinationProvider getLuceneAnalyzerCombinationProvider(@N
...
+        final String[] elements = analyzerPerFieldName.strip().split(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS);
 
-        for (int i = 0; i < elements.length / 2; i++) {
-            String fieldName = elements[2 * i];
-            Pair<AnalyzerChooser, AnalyzerChooser> perFieldAnalyzerChooserPair = getAnalyzerChooser(index, elements[2 * i + 1], type);
+        for (String element : elements) {
+            final String[] keyValuePair = element.strip().split(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE);
+            final String fieldName = keyValuePair[0];
+            final String analyzerName = keyValuePair[1];
+            Pair<AnalyzerChooser, AnalyzerChooser> perFieldAnalyzerChooserPair = getAnalyzerChooser(index, analyzerName, type);
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:109]: As written, this parser algorithm would turn the following string:

```
   a:b,  c: d,e :f
```

Into:

```
{"a": "b", "c": " d", "e ", "f"}
```

That is, it strips out whitespace at the beginning of each key and the end of each value, but not at the end of each key and the beginning of each value. Is that intentional?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java [hunk 11]
```diff
@@ -1,2 +1,88 @@
...
+    @Test
+    void testInvalidIndexOptions() {
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:38]: It might also make sense to add test cases where there is whitespace, and then validate that the whitespace is stripped out as expected

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java (hunk 9 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:36]: The 1st assertion is just a positive test case.

Comment (alecgrieser) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:36]: Right, but I meant more, like, tests that validate that, for example, whitespace is handled correctly and that the strings are parsed into the expected sets or maps, etc.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java (hunk 10 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneAnalyzerRegistryImpl.java:109]: Fixed it

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java [hunk 12]
```diff
@@ -44,5 +46,54 @@ public void validate(@Nonnull MetaDataValidator metaDataValidator) {
...
+                throw new MetaDataException("Index " + index.getName() + " has invalid option value for " + optionKey + ": " + analyzerNamePerFieldOption);
+            }
+            final String[] keyValuePairs = analyzerNamePerFieldOption.split(LuceneIndexOptions.DELIMITER_BETWEEN_ELEMENTS);
+            for (int i = 0 ; i < keyValuePairs.length; i++) {
+                final String keyValue = keyValuePairs[i].strip();
+                int firstIndex = keyValue.indexOf(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE);
+                if (keyValue.isEmpty()
+                        || firstIndex < 1 || firstIndex > keyValue.length() - 2
+                        || firstIndex != keyValue.lastIndexOf(LuceneIndexOptions.DELIMITER_BETWEEN_KEY_AND_VALUE)) {
+                    throw new MetaDataException("Index " + index.getName() + " has invalid option value for " + optionKey + ": " + analyzerNamePerFieldOption);
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java:75]: Should this validate that the (stripped) key and value are non-empty?

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        // valid options
+        final Map<String, String> options = ImmutableMap.of(LuceneIndexOptions.LUCENE_ANALYZER_NAME_PER_FIELD_OPTION, " text1 :NGRAM,text2:SYNONYM",
+                LuceneIndexOptions.AUTO_COMPLETE_EXCLUDED_FIELDS, "text2,text3");
+        validateIndexOptions(options);
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:42]: This should probably should be in an `assertDoesNotThrow` assertion to underscore why this is calling `validateIndexOptions`

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java (hunk 12 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: I think empty key or value should cause exception because the check ```keyValue.isEmpty() || firstIndex < 1 || firstIndex > keyValue.length() - 2```

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java (hunk 9 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java:39]: done

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java (hunk 13 above)
Comment (tian-yizuo) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexValidatorTest.java]: Done

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java (hunk 12 above)
Comment (alecgrieser) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexValidator.java]: Oh, I see. It's a little complicated, but I think you're right: when the `keyValue` is stripped, if there's a completely blank key or value, then either `firstIndex` will be < 1 (for a blank key) or `firstIndex` will be > length - 2 (for a blank value), assuming there's only one delimiter. But I had to convince myself that was enough

**Review by alecgrieser — APPROVED** [review]
Approved, pending a successful PRB (which it seems like is a style thing)

**Review by alecgrieser — APPROVED** [review]

**Review by MMcM — APPROVED** [review]
